### PR TITLE
fix fallocate failed on cephfs

### PR DESCRIPTION
**Pull request by s1m0n21**
when i try to seal a sector on my cephfs, it keeps reporting this error: 
```txt
2020-06-02T14:01:38.781+0800    ERROR   sectors add piece:                                                                                                      
    github.com/filecoin-project/storage-fsm.(*Sealing).pledgeSector
        /root/go/pkg/mod/github.com/filecoin-project/storage-fsm@v0.0.0-20200528050623-cdada6e88960/garbage.go:30
  - creating unsealed sector file:
    github.com/filecoin-project/sector-storage/ffiwrapper.(*Sealer).AddPiece
        /root/go/pkg/mod/github.com/filecoin-project/sector-storage@v0.0.0-20200529175241-9df0cdf19326/ffiwrapper/sealer_cgo.go:91
  - fallocate '/storage/lotus-storage-interop-02/unsealed/s-t01481-1':
    github.com/filecoin-project/sector-storage/ffiwrapper.createPartialFile.func1
        /root/go/pkg/mod/github.com/filecoin-project/sector-storage@v0.0.0-20200529175241-9df0cdf19326/ffiwrapper/partialfile.go:67
  - operation not supported
```
but its normal on other machines, then i found that it can be fixed by rewriting the fallocate function

### CODE REVIEW

**Comment by magik6k** [comment]
From what I can tell ceph should support fallocate - https://github.com/ceph/ceph/blob/356c64a3ff159f2f90c28892f145cf8aa53ba5e0/src/client/fuse_ll.cc#L672-L684

Can you check your fuse version and make sure it's >= 2.9

File: ffiwrapper/partialfile.go [hunk 1]
```diff
@@ -260,3 +259,27 @@ func pieceRun(offset storiface.PaddedByteIndex, size abi.PaddedPieceSize) rleplu
 
 	return &rlepluslazy.RunSliceIterator{Runs: runs}
 }
+
+func fallocate(f *os.File, offset, length int64) error {
```
Comment (magik6k) [inline ffiwrapper/partialfile.go:263]: I wouldn't call this fallocate as it doesn't reserve any space on disk.

**Comment by s1m0n21** [comment]
I tried the kernel driver mount and ceph-fuse mount, both are the same error, and my ceph-fuse version is 
```text
ceph version 12.2.12 (1436006594665279fe734b4c15d7e08c13ebd777) luminous (stable)
```

File: ffiwrapper/partialfile.go [hunk 2]
```diff
@@ -260,3 +259,27 @@ func pieceRun(offset storiface.PaddedByteIndex, size abi.PaddedPieceSize) rleplu
...
+		if length < n {
+			n = length
+		}
+
+		if _, err := f.Write(buf[:n]); err != nil {
+			return err
+		}
+
+		length -= n
+	}
```
Comment (magik6k) [inline ffiwrapper/partialfile.go:282]: We don't need to zero the file, I'm assuming that most filesystems wouldn't leak previous data through files with holes.

Aaand even if they are not zeroed, partialfile should prevent reading these areas

File: ffiwrapper/partialfile.go (hunk 1 above)
Comment (magik6k) [reply ffiwrapper/partialfile.go:263]: (just `createEmpty` would be fine)

**Comment by mostcute** [comment]
ceph worked good on my machine,have not met the err

**Comment by s1m0n21** [comment]
that's weird

**Pull request closed without merging**